### PR TITLE
Hide exclude entries from package buildresults

### PR DESCRIPTION
**Pull request by DavidKang**
By default, we are not showing those entries with state='excluded' as OSC does. 

### Without excluded results
screenshot from 2018-04-04 13-13-46

### With excluded results

buildresult

### When all the results have state excluded

screenshot from 2018-04-04 16-40-40

### CODE REVIEW

**Comment by coolo** [comment]
there will be people looking for reasons their package doesn't build - so what do you display if there are nothing but excludes? (to check put ExclusiveArch: goldenage in the spec file :)

**Comment by DavidKang** [comment]
Will show something like this: 

screenshot from 2018-04-04 15-12-09

**Comment by coolo** [comment]
looks a bit sad, but gives the message :)

File: src/api/app/models/package.rb [hunk 1]
```diff
@@ -939,7 +944,7 @@ def buildresults(prj = project)
       end
     end
 
-    local_build_results
+    return local_build_results, excluded_number
```
Comment (ChrisBr) [inline src/api/app/models/package.rb]: I don't really like this approach :cry: And I also think that it might be better to move this function to a dedicated model now.

What about something like this:
https://gist.github.com/ChrisBr/4e7de315410ea8ccf8cdc149eb210d3e
(just quick and dirty, highly untested. Just to get my idea :smile: )

tl;dr: it would allow sth like:
```
local_build_results = LocalBuildResult::ForPackage.new(package)
local_build_results.all
local_build_results.count(excluded: true)
local_build_results.count
```

It would also make testing easier.

We can brainstorm on that together tomorrow ...

File: src/api/spec/models/package_spec.rb [hunk 2]
```diff
@@ -310,105 +268,9 @@
...
-      it { expect(results_test_package.last.repository).to eq('openSUSE_Tumbleweed') }
-      it { expect(results_test_package.last.architecture).to eq('x86_64') }
-      it { expect(results_test_package.last.code).to eq('building') }
-      it { expect(results_test_package.last.state).to eq('building') }
-      it { expect(results_test_package.last.details).to be_nil }
-
-      it { expect(results_test_package_source.length).to eq(3) }
-      it { expect(results_test_package_image.length).to eq(4) }
+  context '#buildresult' do
+    it 'retrun an object with class LocalBuildResult::ForPackage' do
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: *returns

File: src/api/spec/models/package_spec.rb [hunk 3]
```diff
@@ -20,48 +20,6 @@
   let(:other_user) { create(:confirmed_user, login: 'other_user') }
   let(:other_user2) { create(:confirmed_user, login: 'other_user2') }
   let(:other_user3) { create(:confirmed_user, login: 'other_user3') }
```
Comment (bgeuken) [inline src/api/spec/models/package_spec.rb]: Please add a note to this commit that tells why this test got (re)moved:-)

File: src/api/app/models/local_build_result/for_package.rb [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    attr_accessor :package, :project, :show_excluded
+    attr_reader :excluded_results_count
+
+    def all
+      local_build_results = {}
+      self.excluded_results_count = 0
+      backend_build_result.each do |result|
+        result.elements('status').each do |status|
+          local_build_results[status['package']] ||= []
+          if !show_excluded && excluded?(status['code'])
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: You could move ``show_excluded`` as well in the ``excluded?`` function.

File: src/api/app/models/local_build_result/for_package.rb [hunk 5]
```diff
@@ -0,0 +1,41 @@
+class LocalBuildResult
+  class ForPackage
+    include ActiveModel::Model
+    attr_accessor :package, :project, :show_excluded
+    attr_reader :excluded_results_count
+
+    def all
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: Hmm :thinking: What I don't like now is that you always have to call ``all`` first so that you get the correct result for ``excluded_results_count``.

What about having e.g. a callback (http://api.rubyonrails.org/classes/ActiveModel/Callbacks.html) which does everything what is in ``all`` and you store the result in an instance variable instead.

What I mean is for instance:
Move the content of all to a ``reload`` function which gets called by an after initialize callback. The result in of reload you store in the instance variables ``excluded_results_count`` and ``all`` (don't like this name for a variable, maybe you come up with a better name)...

Sorry, sometimes you have to see everything altogether :cry:

File: src/api/app/models/local_build_result/for_package.rb [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+      local_build_results = {}
+      self.excluded_results_count = 0
+      backend_build_result.each do |result|
+        result.elements('status').each do |status|
+          local_build_results[status['package']] ||= []
+          if !show_excluded && excluded?(status['code'])
+            self.excluded_results_count += 1
+            next
+          end
+          local_build_results[status['package']] << LocalBuildResult.new(repository: result['repository'],
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: You could move that to a helper method like ``build_result(result, status)``

File: src/api/app/models/package.rb [hunk 7]
```diff
@@ -924,22 +924,8 @@ def services
...
-        local_build_results[status['package']] << LocalBuildResult.new(repository: result['repository'],
-                                                                       architecture: result['arch'],
-                                                                       code: status['code'],
-                                                                       state: result['state'],
-                                                                       details: status['details'])
-      end
-    end
-
-    local_build_results
+  def buildresult(prj = project, show_excluded = false)
```
Comment (ChrisBr) [inline src/api/app/models/package.rb]: You need this wrapper method because there are too many other uses, right (otherwise I would get rid of it ... But it is a nitpick, we can do this later on as well)

File: src/api/app/views/webui/package/_buildstatus.html.haml [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+    = check_box_tag :show_excluded, true, @show_excluded
+    - label_message = @excluded_results_count.zero? ? 'Hide' :  "Show #{@excluded_results_count}"
+    = label_tag :show_excluded, "#{label_message} excluded results"
+
+- @buildresults.each_pair do |package, results|
+  %h3= package
+  #package-buildstatus
+    %table
+      - if results.present?
+        - previous_repo = nil
```
Comment (ChrisBr) [inline src/api/app/views/webui/package/_buildstatus.html.haml]: Out of curiosity: why do we need to save the previous repository and compare it with the current?

File: src/api/spec/models/package_spec.rb (hunk 3 above)
Comment (DavidKang) [reply src/api/spec/models/package_spec.rb]: done

File: src/api/app/views/webui/package/_buildstatus.html.haml (hunk 8 above)
Comment (DavidKang) [reply src/api/app/views/webui/package/_buildstatus.html.haml]: IIRC, is for not repeat the repository in each architecture.

**Comment by DavidKang** [comment]
@ChrisBr and @bgeuken could you review it again please :smile_cat:

**Comment by ChrisBr** [comment]
review app is available here: http://10.86.0.147/davidkang-sprint36featurep4-hide-exclude-from-package-build

File: src/api/app/models/local_build_result/for_package.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+
+    def initialize(attributes = {})
+      super
+      @excluded_counter = 0
+      buildresults
+    end
+
+    private
+
+    attr_writer :excluded_results_count, :results
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: excluded_results_count is not used, is it?

File: src/api/app/models/local_build_result/for_package.rb [hunk 10]
```diff
@@ -0,0 +1,45 @@
+class LocalBuildResult
+  class ForPackage
+    include ActiveModel::Model
+    attr_accessor :package, :project, :show_all
+    attr_reader :excluded_counter, :results
+
+    def initialize(attributes = {})
+      super
+      @excluded_counter = 0
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: Why is it an instance variable instead of getter/setter?

File: src/api/app/models/local_build_result/for_package.rb [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+      buildresults
+    end
+
+    private
+
+    attr_writer :excluded_results_count, :results
+
+    def buildresults
+      @results = {}
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: You could do this in the getter, of course you don't need the attr_reader anymore:

```
def results
  @results || {}
end
```

File: src/api/app/models/local_build_result/for_package.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      backend_build_result.each do |result|
+        result.elements('status').each do |status|
+          @results[status['package']] ||= []
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: You could use the setter method here with ``self.results``

File: src/api/app/models/local_build_result/for_package.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+          if excluded?(status['code'])
+            @excluded_counter += 1
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: same here

File: src/api/app/models/local_build_result/for_package.rb [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+            next
+          end
+          add_local_build_result(result, status)
+        end
+      end
+    end
+
+    def add_local_build_result(result, status)
+      @results[status['package']] << LocalBuildResult.new(repository: result['repository'], architecture: result['arch'],
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: [nitpick] I would assign it in the buildresult method
Also you can access the variable with the setter method self.result

File: src/api/app/models/local_build_result/for_package.rb [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+      buildresults
+    end
+
+    private
+
+    attr_writer :excluded_counter, :results
+
+    def buildresults
+      self.results = {}
+      self.excluded_counter = 0
```
Comment (ChrisBr) [inline src/api/app/models/local_build_result/for_package.rb]: [nitpick] You can do that in the getter
```
def results
  @results || {}
end

def excluded_counter
  @excluded_counter || 0
end
```

**Comment by bgeuken** [comment]
Nice work :smiley:
